### PR TITLE
Decoder mixed errors test

### LINKED ISSUE #209: Single integration test exercising multiple message edge cases

### PR DESCRIPTION
**Pull request by jklukas**
Writing this test uncovered a few bugs and some additional testing nuance, so there are several things going on. I've split some of those changes out in separate commits, but the final commit that adds the test is still rather large and contains some additional changes. I'll add inline comments to explain more of the supporting changes.

Closes #209

### CODE REVIEW

File: ingestion-beam/src/main/java/com/mozilla/telemetry/Decoder.java [hunk 1]
```diff
@@ -34,42 +37,46 @@ public static void main(String[] args) {
 
     final DecoderOptions.Parsed options = DecoderOptions.parseDecoderOptions(
         PipelineOptionsFactory.fromArgs(args).withValidation().as(DecoderOptions.class));
-
     final Pipeline pipeline = Pipeline.create(options);
-    final PTransform<PCollection<PubsubMessage>, ? extends POutput> errorOutput = options
-        .getErrorOutputType().write(options);
+    final List<PCollection<PubsubMessage>> errorCollections = new ArrayList<>();
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/Decoder.java]: One of the bugs discovered here is that we were creating multiple transforms for each step's error output, but the configurations are identical, so they create potentially identical filenames and overwrite each other. We now collect error collections in a list and combine to one error output transform.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/Sink.java [hunk 2]
```diff
@@ -27,17 +30,21 @@ public static void main(String[] args) {
 
     final SinkOptions.Parsed options = SinkOptions.parseSinkOptions(
         PipelineOptionsFactory.fromArgs(args).withValidation().as(SinkOptions.class));
-
     final Pipeline pipeline = Pipeline.create(options);
-    final PTransform<PCollection<PubsubMessage>, ? extends POutput> errorOutput = options
-        .getErrorOutputType().write(options);
+    final List<PCollection<PubsubMessage>> errorCollections = new ArrayList<>();
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/Sink.java]: Same error output change in Sink as in Decoder.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java [hunk 3]
```diff
@@ -55,9 +55,9 @@
 
   @Description("URI of a redis server that will be used for deduplication")
   @Validation.Required
-  ValueProvider<URI> getRedisUri();
+  ValueProvider<String> getRedisUri();
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecoderOptions.java]: It looks like Beam was passing this to Jackson; Jackson was expecting a JSON blob that it could deserialize into a URI, which isn't what we want.

So I changed this to String and we now have a separate parsing step.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/options/ErrorOutputType.java [hunk 4]
```diff
@@ -5,19 +5,24 @@
...
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessageWithAttributesCoder;
+import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.values.PCollection;
 import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.TypeDescriptor;
 
 /**
- * Enumeration of error output types that each provide a {@code write} method.
+ * Enumeration of error output types that each provide a {@code writeFailures} method.
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/options/ErrorOutputType.java]: We now have a concrete `write` method that calls the abstract `writeFailures` method. This allows us to consolidate logic for the new `includeStackTraces` configuration option.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java [hunk 5]
```diff
@@ -20,25 +27,39 @@
...
   public TemporaryFolder outputFolder = new TemporaryFolder();
 
-  private String outputPath;
-
-  @Before
-  public void initialize() {
-    outputPath = outputFolder.getRoot().getAbsolutePath();
-  }
+  @Rule
+  public final RedisServer redis = new RedisServer();
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java]: Since RedisServer is now needed in two places, I implemented a junit `Rule` for it that gets used here and in DecoderTest

File: ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java [hunk 6]
```diff
@@ -20,25 +27,39 @@
...
+  }
+
+  /** Reset the ObjectMapper configurations we changed. */
+  @AfterClass
+  public static void tearDown() throws Exception {
+    Field mapperField = Json.class.getDeclaredField("MAPPER");
+    mapperField.setAccessible(true);
+    ObjectMapper mapper = (ObjectMapper) mapperField.get(null);
+    mapper.disable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS);
+  }
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java]: This reflection business is a bit unfortunate, but I think it's good to keep the mapper private. In the future, it would be nice to build more robust testing utilities that can compare PubsubMessages rather than strings and could avoid this need.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderTest.java [hunk 7]
```diff
@@ -1,323 +0,0 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderTest.java]: This file was supposed to be deleted in a previous PR where I broke this into separate classes. I somehow screwed that up and this file got included again before merging. Cleaning that up here.

File: ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/DeduplicateTest.java [hunk 8]
```diff
@@ -23,87 +24,78 @@
...
     Integer ttlSeconds = parsedOptions.getDeduplicateExpireSeconds().get();
-
-    int redisPort = new redis.embedded.ports.EphemeralPortProvider().next();
-    // create testing redis server
-    final RedisServer redis = RedisServer.builder().port(redisPort).setting("bind 127.0.0.1")
-        .build();
-    final URI redisUri = URI.create("redis://localhost:" + redisPort);
-    redis.start();
-
-    try {
```
Comment (jklukas) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/decoder/DeduplicateTest.java]: Changes in this block are just indentation because we no longer need the try/catch with the RedisServer rule

File: ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java (hunk 6 above)
Comment (jklukas) [reply ingestion-beam/src/test/java/com/mozilla/telemetry/DecoderMainTest.java]: Wrote up https://github.com/mozilla/gcp-ingestion/issues/234
